hip/test/components/cooperative_groups.hip.cpp

```diff
@@ -0,0 +1,204 @@
+    if (threadIdx.x == 0) {
+        *success &= full;
+    }
+}
+
+
+__global__ void cg_shuffle(bool *s)
+{
+    auto group = tiled_partition<config::warp_size>(this_thread_block());
+    auto i = int(group.thread_rank());
+    test_assert(s, group.shfl_up(i, 1) == max(0, i - 1));
+    test_assert(s, group.shfl_down(i, 1) == min(i + 1, config::warp_size - 1));
+    test_assert(s, group.shfl(i, 0) == 0);
+}
+
+TEST_F(CooperativeGroups, Shuffle) { test(cg_shuffle); }
+
+
+__global__ void cg_all(bool *s)
+{
+    auto group = tiled_partition<config::warp_size>(this_thread_block());
+    test_assert(s, group.all(true));
+    test_assert(s, !group.all(false));
+    test_assert(s, !group.all(threadIdx.x < 13));
+}
+
+TEST_F(CooperativeGroups, All) { test(cg_all); }
+
+
+__global__ void cg_any(bool *s)
+{
+    auto group = tiled_partition<config::warp_size>(this_thread_block());
+    test_assert(s, group.any(true));
+    test_assert(s, group.any(threadIdx.x == 0));
+    test_assert(s, !group.any(false));
+}
+
+TEST_F(CooperativeGroups, Any) { test(cg_any); }
+
+
+__global__ void cg_ballot(bool *s)
+{
+    auto group = tiled_partition<config::warp_size>(this_thread_block());
+    test_assert(s, group.ballot(false) == 0);
+    test_assert(s, group.ballot(true) == ~config::lane_mask_type{});
+    test_assert(s, group.ballot(threadIdx.x < 4) == 0xf);
+}
+
+TEST_F(CooperativeGroups, Ballot) { test(cg_ballot); }
+
+
+__global__ void cg_subwarp_shuffle(bool *s)
+{
+    auto group = tiled_partition<subwarp_size>(this_thread_block());
+    auto i = int(group.thread_rank());
+    test_assert(s, group.shfl_up(i, 1) == max(i - 1, 0));
+    test_assert(s, group.shfl_down(i, 1) == min(i + 1, subwarp_size - 1));
+    auto group_base = threadIdx.x / subwarp_size * subwarp_size;
+    test_assert(s, group.shfl(int(threadIdx.x), 0) == group_base);
+}
+
+TEST_F(CooperativeGroups, SubwarpShuffle) { test(cg_subwarp_shuffle); }
+
+
+__global__ void cg_subwarp_all(bool *s)
+{
+    auto grp = threadIdx.x / subwarp_size;
+    bool test_grp = grp == 1;
+    auto i = threadIdx.x % subwarp_size;
+    // only test with test_grp, the other threads run 'interference'
+    auto group = tiled_partition<subwarp_size>(this_thread_block());
+    test_assert(s, !test_grp || group.all(test_grp));
+    test_assert(s, !test_grp || !group.all(!test_grp));
+    test_assert(s, !test_grp || !group.all(i < subwarp_size - 3 || !test_grp));
+}
+
+TEST_F(CooperativeGroups, SubwarpAll) { test(cg_subwarp_all); }
+
+
+__global__ void cg_subwarp_any(bool *s)
+{
+    auto grp = threadIdx.x / subwarp_size;
+    bool test_grp = grp == 1;
+    // only test with test_grp, the other threads run 'interference'
+    auto group = tiled_partition<subwarp_size>(this_thread_block());
+    auto i = group.thread_rank();
+    test_assert(s, !test_grp || group.any(test_grp));
+    test_assert(s, !test_grp || group.any(test_grp && i == 1));
+    test_assert(s, !test_grp || !group.any(!test_grp));
+}
```

Comment:
for any, all, and ballot, I would like to test the following things. (maybe testing on ballot is enough)
1. use warp_size on kernel, and if (idx < subwarp_size) -> evaluate group and test
2. use subwarp_size on kernel, evaluate group and test
3. use warp_size on kernel, and use if - else -> evaluate different operation in different block